# 1.Analysis of Steam database for historical trends:

We will start by examining historical data from the Steam marketplace, focusing on the 10 most popular games labelled "Automobile Sim" from 2013 to 2023. The data comes from .csv files provided by the SteamDB website. 
Through this analysis, we aim to discover trends, patterns and fluctuations in the popularity of racing simulation games over the last decade.

we got .csv files fromn the SteamDB platforms importated locally.

## Import libraries


In [147]:
import os
import numpy as np
import pandas as pd

import plotly.express as px
import plotly.graph_objects as go

## Load & concatenate all data sets
all .csv files have the same structure, same variables.

In [148]:
# Set the path of the folder containing the CSV files
folder_path = '/Users/macbook/Dropbox/Mac/Documents/Pro/Data_Analyst/simracing-players/data/raw'

# Initialize an empty dictionary to store DataFrames
dfs = {}

# Iterate through each file in the folder
for file in os.listdir(folder_path):
   # Check if the current file is a CSV file
   if file.endswith('.csv'):
       # Extract game name from filename and create an empty DataFrame
       game_name = os.path.splitext(file)[0]
       df = pd.DataFrame()

       # Read the current CSV file using Pandas and assign it to the DataFrame
       df = pd.read_csv(os.path.join(folder_path, file))

       # Add 'game' column with game name for each DataFrame
       df['game'] = game_name

       # Store the current DataFrame in the dictionary using its game name as a key
       dfs[game_name] = df

In [149]:
merged_df = pd.concat(dfs.values(), ignore_index=True)

### Checking the load & merge

In [150]:
merged_df.head(20)

DateTime  Players  Average Players  Twitch Viewers  \
0   2013-10-01 00:00:00      8.0              NaN             NaN   
1   2013-10-02 00:00:00      NaN              NaN             NaN   
2   2013-10-03 00:00:00      NaN              NaN             NaN   
3   2013-10-04 00:00:00      NaN              NaN             NaN   
4   2013-10-05 00:00:00      NaN              NaN             NaN   
5   2013-10-06 00:00:00      NaN              NaN             NaN   
6   2013-10-07 00:00:00      NaN              NaN             NaN   
7   2013-10-08 00:00:00      NaN              NaN             NaN   
8   2013-10-09 00:00:00      NaN              NaN             NaN   
9   2013-10-10 00:00:00      NaN              NaN             NaN   
10  2013-10-11 00:00:00      NaN              NaN             NaN   
11  2013-10-12 00:00:00      NaN              NaN             NaN   
12  2013-10-13 00:00:00      NaN              NaN             NaN   
13  2013-10-14 00:00:00      NaN              NaN             NaN   
14  2013-10-15 00:00:00      NaN              NaN             NaN   
15  2013-10-16 00:00:00      NaN              NaN             NaN   
16  2013-10-17 00:00:00      NaN              NaN             NaN   
17  2013-10-18 00:00:00      NaN              NaN             NaN   
18  2013-10-19 00:00:00      NaN              NaN             NaN   
19  2013-10-20 00:00:00      NaN              NaN             NaN   

             game  
0   Assetto_corsa  
1   Assetto_corsa  
2   Assetto_corsa  
3   Assetto_corsa  
4   Assetto_corsa  
5   Assetto_corsa  
6   Assetto_corsa  
7   Assetto_corsa  
8   Assetto_corsa  
9   Assetto_corsa  
10  Assetto_corsa  
11  Assetto_corsa  
12  Assetto_corsa  
13  Assetto_corsa  
14  Assetto_corsa  
15  Assetto_corsa  
16  Assetto_corsa  
17  Assetto_corsa  
18  Assetto_corsa  
19  Assetto_corsa

In [151]:
merged_df.tail(20)

DateTime  Players  Average Players  Twitch Viewers  \
40934  2024-01-09 10:40:00    699.0              NaN            22.0   
40935  2024-01-09 10:50:00    692.0              NaN            22.0   
40936  2024-01-09 11:00:00    712.0            966.0            11.0   
40937  2024-01-09 11:10:00    729.0              NaN            11.0   
40938  2024-01-09 11:20:00    753.0              NaN            11.0   
40939  2024-01-09 11:30:00    776.0              NaN            11.0   
40940  2024-01-09 11:40:00    800.0              NaN            11.0   
40941  2024-01-09 11:50:00    828.0              NaN            11.0   
40942  2024-01-09 12:00:00    843.0            966.0            14.0   
40943  2024-01-09 12:10:00    852.0              NaN            14.0   
40944  2024-01-09 12:20:00    857.0              NaN            14.0   
40945  2024-01-09 12:30:00    859.0              NaN            14.0   
40946  2024-01-09 12:40:00    881.0              NaN            14.0   
40947  2024-01-09 12:50:00    895.0              NaN            14.0   
40948  2024-01-09 13:00:00    908.0            960.0            18.0   
40949  2024-01-09 13:10:00    913.0              NaN            18.0   
40950  2024-01-09 13:20:00    923.0              NaN            18.0   
40951  2024-01-09 13:30:00      NaN              NaN            18.0   
40952  2024-01-09 13:40:00      NaN              NaN            18.0   
40953  2024-01-09 13:50:00      NaN              NaN            18.0   

                  game  
40934  DiRT_rally_2.0_  
40935  DiRT_rally_2.0_  
40936  DiRT_rally_2.0_  
40937  DiRT_rally_2.0_  
40938  DiRT_rally_2.0_  
40939  DiRT_rally_2.0_  
40940  DiRT_rally_2.0_  
40941  DiRT_rally_2.0_  
40942  DiRT_rally_2.0_  
40943  DiRT_rally_2.0_  
40944  DiRT_rally_2.0_  
40945  DiRT_rally_2.0_  
40946  DiRT_rally_2.0_  
40947  DiRT_rally_2.0_  
40948  DiRT_rally_2.0_  
40949  DiRT_rally_2.0_  
40950  DiRT_rally_2.0_  
40951  DiRT_rally_2.0_  
40952  DiRT_rally_2.0_  
40953  DiRT_rally_2.0_

we see that in 2013, we had one records of player per day. In our most recent records, it was done every 10 mins.

In [152]:
merged_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40954 entries, 0 to 40953
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   DateTime         40954 non-null  object 
 1   Players          38509 non-null  float64
 2   Average Players  11811 non-null  float64
 3   Twitch Viewers   36883 non-null  float64
 4   game             40954 non-null  object 
dtypes: float64(3), object(2)
memory usage: 1.6+ MB


We need to change the data type of datetime.

In [153]:
round(merged_df.describe())

Players  Average Players  Twitch Viewers
count  38509.0          11811.0         36883.0
mean    8678.0           8268.0          1177.0
std    10489.0           8194.0          3608.0
min        1.0            265.0             0.0
25%     1805.0           2209.0            57.0
50%     4518.0           6389.0           290.0
75%    11379.0          10886.0           930.0
max    81096.0          44392.0        127965.0

## Data preparation & cleaning

Let's rename the column headers

In [154]:
print(merged_df.columns)

Index(['DateTime', 'Players', 'Average Players', 'Twitch Viewers', 'game'], dtype='object')


In [155]:
# rename column headers
merged_df.columns = ['datetime', 'players', 'average_players', 'twitch_viewers', 'game']

In [156]:
merged_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40954 entries, 0 to 40953
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   datetime         40954 non-null  object 
 1   players          38509 non-null  float64
 2   average_players  11811 non-null  float64
 3   twitch_viewers   36883 non-null  float64
 4   game             40954 non-null  object 
dtypes: float64(3), object(2)
memory usage: 1.6+ MB


In [157]:
merged_df['datetime'] = merged_df['datetime'].astype('datetime64[s]')
merged_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40954 entries, 0 to 40953
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype        
---  ------           --------------  -----        
 0   datetime         40954 non-null  datetime64[s]
 1   players          38509 non-null  float64      
 2   average_players  11811 non-null  float64      
 3   twitch_viewers   36883 non-null  float64      
 4   game             40954 non-null  object       
dtypes: datetime64[s](1), float64(3), object(1)
memory usage: 1.6+ MB


In [158]:
#check for mixed types
for col in merged_df.columns.tolist():
  weird = (merged_df[[col]].map(type) != merged_df[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (merged_df[weird]) > 0:
    print (col)

In [159]:
merged_df.tail()

datetime  players  average_players  twitch_viewers  \
40949 2024-01-09 13:10:00    913.0              NaN            18.0   
40950 2024-01-09 13:20:00    923.0              NaN            18.0   
40951 2024-01-09 13:30:00      NaN              NaN            18.0   
40952 2024-01-09 13:40:00      NaN              NaN            18.0   
40953 2024-01-09 13:50:00      NaN              NaN            18.0   

                  game  
40949  DiRT_rally_2.0_  
40950  DiRT_rally_2.0_  
40951  DiRT_rally_2.0_  
40952  DiRT_rally_2.0_  
40953  DiRT_rally_2.0_

In [160]:
# Checking the null values
print(merged_df.isnull().sum())


datetime               0
players             2445
average_players    29143
twitch_viewers      4071
game                   0
dtype: int64


In [161]:
# Check for duplicates
print(merged_df.duplicated().sum())


0


In [162]:
# have a look on Null values 
null = merged_df[merged_df[['players', 'average_players', 'twitch_viewers']].isnull().all(axis=1)]
null

datetime  players  average_players  twitch_viewers             game
1     2013-10-02      NaN              NaN             NaN    Assetto_corsa
2     2013-10-03      NaN              NaN             NaN    Assetto_corsa
3     2013-10-04      NaN              NaN             NaN    Assetto_corsa
4     2013-10-05      NaN              NaN             NaN    Assetto_corsa
5     2013-10-06      NaN              NaN             NaN    Assetto_corsa
...          ...      ...              ...             ...              ...
37494 2018-10-15      NaN              NaN             NaN  DiRT_rally_2.0_
37495 2018-10-16      NaN              NaN             NaN  DiRT_rally_2.0_
37496 2018-10-17      NaN              NaN             NaN  DiRT_rally_2.0_
37497 2018-10-18      NaN              NaN             NaN  DiRT_rally_2.0_
37498 2018-10-19      NaN              NaN             NaN  DiRT_rally_2.0_

[1131 rows x 5 columns]

Let's delete all the records with null values in ALL columns except game

In [163]:
merged_df = merged_df.dropna(subset=['players', 'average_players', 'twitch_viewers'], how='all')
merged_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 39823 entries, 0 to 40953
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype        
---  ------           --------------  -----        
 0   datetime         39823 non-null  datetime64[s]
 1   players          38509 non-null  float64      
 2   average_players  11811 non-null  float64      
 3   twitch_viewers   36883 non-null  float64      
 4   game             39823 non-null  object       
dtypes: datetime64[s](1), float64(3), object(1)
memory usage: 1.8+ MB


players, average players & twitch viewers are individuals so the data type must be integer.

In [164]:
print(merged_df.isnull().sum())

datetime               0
players             1314
average_players    28012
twitch_viewers      2940
game                   0
dtype: int64


we need to replace NA values by 0 before changing the data type.

In [165]:
# replace NA values by 0
merged_df[['players', 'average_players', 'twitch_viewers']] = merged_df[['players', 'average_players', 'twitch_viewers']].fillna(0)
print(merged_df.isnull().sum())

datetime           0
players            0
average_players    0
twitch_viewers     0
game               0
dtype: int64


In [166]:
# Change data type from float to integer
merged_df[['players', 'average_players', 'twitch_viewers']] = merged_df[['players', 'average_players', 'twitch_viewers']].astype(int)

# Display changes
print(merged_df.info())

<class 'pandas.core.frame.DataFrame'>
Index: 39823 entries, 0 to 40953
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype        
---  ------           --------------  -----        
 0   datetime         39823 non-null  datetime64[s]
 1   players          39823 non-null  int64        
 2   average_players  39823 non-null  int64        
 3   twitch_viewers   39823 non-null  int64        
 4   game             39823 non-null  object       
dtypes: datetime64[s](1), int64(3), object(1)
memory usage: 1.8+ MB
None


In [167]:
merged_df['game'].unique()

array(['Assetto_corsa', 'Forza Horizon 5 Steam Charts',
       'Forza Horizon 4 Steam Charts', 'Assetto_corsa_competizione',
       'ATS Price', 'BeamNG', 'Euro_truck_2',
       'Automobilista 2 Steam Charts', 'CarX_drift_racing_online',
       'DiRT_rally_2.0_'], dtype=object)

Let's modify the name of game records to keep consistency

In [168]:
# Normalize records from 'game' field
merged_df['game'] = merged_df['game'].replace({
    'Forza Horizon 5 Steam Charts': 'Forza_horizon_5',
    'Forza Horizon 4 Steam Charts':'Forza_horizon_4',
    'ATS Price': 'American_truck_simulator',
    'Automobilista 2 Steam Charts': 'Automobilista_2',
    'DiRT_rally_2.0_': 'Dirt_rally_2.0' 
})
# checking changes
merged_df['game'].unique()

array(['Assetto_corsa', 'Forza_horizon_5', 'Forza_horizon_4',
       'Assetto_corsa_competizione', 'American_truck_simulator', 'BeamNG',
       'Euro_truck_2', 'Automobilista_2', 'CarX_drift_racing_online',
       'Dirt_rally_2.0'], dtype=object)

## Exploratary analysis & visualisation

In [169]:
round(merged_df.describe())


datetime  players  average_players  twitch_viewers
count                39823  39823.0          39823.0         39823.0
mean   2021-09-15 18:51:56   8392.0           2452.0          1090.0
min    2013-01-01 00:00:00      0.0              0.0             0.0
25%    2019-11-27 00:00:00   1604.0              0.0            35.0
50%    2022-10-26 00:00:00   4271.0              0.0           240.0
75%    2023-12-31 19:30:00  10992.0           1266.0           824.0
max    2024-02-22 16:30:00  81096.0          44392.0        127965.0
std                    NaN  10431.0           5846.0          3485.0

we have records from 1st january of 2013 until february 2024.
We will keep only records until 31st december of 2023.

In [170]:
# remove records from 2024
df_2013_2023 = merged_df[merged_df['datetime'] <= '2023-12-31']
round(df_2013_2023.describe())


datetime  players  average_players  twitch_viewers
count                29501  29501.0          29501.0         29501.0
mean   2020-11-23 04:06:41   8569.0           2617.0          1322.0
min    2013-01-01 00:00:00      0.0              0.0             0.0
25%    2019-01-08 00:00:00   1454.0              0.0            33.0
50%    2021-05-26 00:00:00   3995.0              0.0           275.0
75%    2023-06-05 00:00:00  11325.0           1554.0          1137.0
max    2023-12-31 00:00:00  81096.0          44392.0        127965.0
std                    NaN  10923.0           6126.0          3980.0

### Overview of records between 2013-2023:

In [171]:
# plot representing records in the 2013-2023 time period
fig = px.histogram(df_2013_2023, x='datetime', template='plotly_dark')

# update shape of the plot
fig.update_layout(
    xaxis=dict(showgrid=False),
    yaxis=dict(showgrid=False)
)
fig.show()

We have a lot of records the last month of 2023. Let's create new 'month' & 'year' columns

In [172]:
# Extract month from the 'datetime' column
df_2013_2023.loc[:, 'month'] = df_2013_2023['datetime'].dt.month

# Extract year from the 'datetime' column
df_2013_2023.loc[:, 'year'] = df_2013_2023['datetime'].dt.year

# Filter the DataFrame to keep records for the year 2023
df_2023 = df_2013_2023[df_2013_2023['year'] == 2023]


/var/folders/xh/_q4grmbx5wv2fs3fc3vrkwmc0000gn/T/ipykernel_34121/3136023133.py:2: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

/var/folders/xh/_q4grmbx5wv2fs3fc3vrkwmc0000gn/T/ipykernel_34121/3136023133.py:5: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



Visualize the count of records of year 2023: 

In [173]:
# Create the histogram with month of 2023
hist_2023 = px.histogram(df_2023, x='month', template='plotly_dark')

# Update layout to remove gridlines
hist_2023.update_layout(
    xaxis=dict(showgrid=False), 
    yaxis=dict(showgrid=False)
)

hist_2023.show()


In [174]:
# defining a variable dec for December month
dec = df_2023[df_2023['month'].isin([12])]
dec.head(20)

datetime  players  average_players  twitch_viewers  \
3713 2023-12-01 00:00:00    14820             8643            2414   
3714 2023-12-02 00:00:00    16840            10592             889   
3715 2023-12-03 00:00:00    15706            10297             898   
3716 2023-12-04 00:00:00    13087             7331            1628   
3717 2023-12-05 00:00:00    12754             7335            5271   
3718 2023-12-06 00:00:00    13015             7565            1405   
3719 2023-12-07 00:00:00    13951             7870             617   
3720 2023-12-08 00:00:00    15604             9030             606   
3721 2023-12-09 00:00:00    16517            10962            2677   
3722 2023-12-10 00:00:00    16495            10848            2761   
3723 2023-12-11 00:00:00    14004             8031            1932   
3724 2023-12-12 00:00:00     7514             8191             209   
3725 2023-12-12 01:00:00     7196             8182             206   
3726 2023-12-12 02:00:00     7026             8167             195   
3727 2023-12-12 03:00:00     6560             8149             140   
3728 2023-12-12 04:00:00     5718             8121              82   
3729 2023-12-12 05:00:00     4891             8101             104   
3730 2023-12-12 06:00:00     4512             8094              98   
3731 2023-12-12 07:00:00     4318             8083              79   
3732 2023-12-12 08:00:00     4405             8072              75   

               game  month  year  
3713  Assetto_corsa     12  2023  
3714  Assetto_corsa     12  2023  
3715  Assetto_corsa     12  2023  
3716  Assetto_corsa     12  2023  
3717  Assetto_corsa     12  2023  
3718  Assetto_corsa     12  2023  
3719  Assetto_corsa     12  2023  
3720  Assetto_corsa     12  2023  
3721  Assetto_corsa     12  2023  
3722  Assetto_corsa     12  2023  
3723  Assetto_corsa     12  2023  
3724  Assetto_corsa     12  2023  
3725  Assetto_corsa     12  2023  
3726  Assetto_corsa     12  2023  
3727  Assetto_corsa     12  2023  
3728  Assetto_corsa     12  2023  
3729  Assetto_corsa     12  2023  
3730  Assetto_corsa     12  2023  
3731  Assetto_corsa     12  2023  
3732  Assetto_corsa     12  2023

In [175]:
# groups by game and counts the number of occurrences of each game in the 'datetime' column.
dec.groupby('game').datetime.count()

game
American_truck_simulator        31
Assetto_corsa                  468
Assetto_corsa_competizione     514
Automobilista_2                514
BeamNG                        1016
CarX_drift_racing_online       514
Dirt_rally_2.0                 514
Euro_truck_2                   987
Forza_horizon_4                514
Forza_horizon_5                514
Name: datetime, dtype: int64

We see that except for American_truck_simulator, we have more than 1 record a day.
From december 2023, the records get hourly. to keep consistency, we will resample the data to daily with the records at 00:00:00

### Resample data frame to keep only one record per day


In [176]:
df_2013_2023.info()

<class 'pandas.core.frame.DataFrame'>
Index: 29501 entries, 0 to 39880
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype        
---  ------           --------------  -----        
 0   datetime         29501 non-null  datetime64[s]
 1   players          29501 non-null  int64        
 2   average_players  29501 non-null  int64        
 3   twitch_viewers   29501 non-null  int64        
 4   game             29501 non-null  object       
 5   month            29501 non-null  int32        
 6   year             29501 non-null  int32        
dtypes: datetime64[s](1), int32(2), int64(3), object(1)
memory usage: 1.6+ MB


In [177]:
# Set datetime as index
df_2013_2023.set_index(['datetime'], inplace=True)
df_2013_2023.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 29501 entries, 2013-10-01 00:00:00 to 2023-12-31 00:00:00
Data columns (total 6 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   players          29501 non-null  int64 
 1   average_players  29501 non-null  int64 
 2   twitch_viewers   29501 non-null  int64 
 3   game             29501 non-null  object
 4   month            29501 non-null  int32 
 5   year             29501 non-null  int32 
dtypes: int32(2), int64(3), object(1)
memory usage: 1.4+ MB


In [178]:
round(df_2013_2023.describe())

players  average_players  twitch_viewers    month     year
count  29501.0          29501.0         29501.0  29501.0  29501.0
mean    8569.0           2617.0          1322.0      8.0   2020.0
std    10923.0           6126.0          3980.0      4.0      3.0
min        0.0              0.0             0.0      1.0   2013.0
25%     1454.0              0.0            33.0      4.0   2019.0
50%     3995.0              0.0           275.0      8.0   2021.0
75%    11325.0           1554.0          1137.0     12.0   2023.0
max    81096.0          44392.0        127965.0     12.0   2023.0

In [179]:
# Group by 'game' before resampling
df_resampled = df_2013_2023.groupby('game').resample('D').first()

# Drop the 'game' column
df_resampled = df_resampled.drop(columns='game')
df_resampled

players  average_players  twitch_viewers  \
game                     datetime                                               
American_truck_simulator 2015-04-19      1.0              0.0             0.0   
                         2015-04-20      NaN              NaN             NaN   
                         2015-04-21      1.0              0.0             0.0   
                         2015-04-22      NaN              NaN             NaN   
                         2015-04-23      NaN              NaN             NaN   
...                                      ...              ...             ...   
Forza_horizon_5          2023-12-27  12773.0          16719.0           420.0   
                         2023-12-28  12726.0          16998.0           397.0   
                         2023-12-29  13925.0          18034.0           737.0   
                         2023-12-30  13610.0          18661.0           463.0   
                         2023-12-31  13948.0          18762.0           395.0   

                                     month    year  
game                     datetime                   
American_truck_simulator 2015-04-19    4.0  2015.0  
                         2015-04-20    NaN     NaN  
                         2015-04-21    4.0  2015.0  
                         2015-04-22    NaN     NaN  
                         2015-04-23    NaN     NaN  
...                                    ...     ...  
Forza_horizon_5          2023-12-27   12.0  2023.0  
                         2023-12-28   12.0  2023.0  
                         2023-12-29   12.0  2023.0  
                         2023-12-30   12.0  2023.0  
                         2023-12-31   12.0  2023.0  

[25356 rows x 5 columns]

In [180]:
# Reset the multi level index
df_resampled = df_resampled.reset_index()

df_resampled.tail(10)

game   datetime  players  average_players  twitch_viewers  \
25346  Forza_horizon_5 2023-12-22   9311.0          14698.0           392.0   
25347  Forza_horizon_5 2023-12-23  10849.0          16555.0           377.0   
25348  Forza_horizon_5 2023-12-24  11568.0          17203.0           500.0   
25349  Forza_horizon_5 2023-12-25  11504.0          16821.0           296.0   
25350  Forza_horizon_5 2023-12-26  12939.0          16750.0           303.0   
25351  Forza_horizon_5 2023-12-27  12773.0          16719.0           420.0   
25352  Forza_horizon_5 2023-12-28  12726.0          16998.0           397.0   
25353  Forza_horizon_5 2023-12-29  13925.0          18034.0           737.0   
25354  Forza_horizon_5 2023-12-30  13610.0          18661.0           463.0   
25355  Forza_horizon_5 2023-12-31  13948.0          18762.0           395.0   

       month    year  
25346   12.0  2023.0  
25347   12.0  2023.0  
25348   12.0  2023.0  
25349   12.0  2023.0  
25350   12.0  2023.0  
25351   12.0  2023.0  
25352   12.0  2023.0  
25353   12.0  2023.0  
25354   12.0  2023.0  
25355   12.0  2023.0

Check now our record counts

In [181]:
df_resampled.describe()

datetime       players  average_players  twitch_viewers  \
count                25356  24225.000000     24225.000000    24225.000000   
mean   2020-01-06 10:39:45   7830.698452      1350.667245     1475.811434   
min    2013-01-01 00:00:00      0.000000         0.000000        0.000000   
25%    2018-03-08 00:00:00   1348.000000         0.000000       39.000000   
50%    2020-06-04 00:00:00   3576.000000         0.000000      337.000000   
75%    2022-04-07 00:00:00   9673.000000         0.000000     1329.000000   
max    2023-12-31 00:00:00  81096.000000     44392.000000   127965.000000   
std                    NaN  10260.238504      4088.188462     4307.802281   

              month          year  
count  24225.000000  24225.000000  
mean       6.719670   2019.710093  
min        1.000000   2013.000000  
25%        4.000000   2018.000000  
50%        7.000000   2020.000000  
75%       10.000000   2022.000000  
max       12.000000   2023.000000  
std        3.445792      2.553853

In [182]:
# checking the number of records in december 2023
df_resampled[(df_resampled['month'] == 12) & (df_resampled['year'] == 2023)].groupby('game')['datetime'].count()


game
American_truck_simulator      31
Assetto_corsa                 31
Assetto_corsa_competizione    31
Automobilista_2               31
BeamNG                        31
CarX_drift_racing_online      31
Dirt_rally_2.0                31
Euro_truck_2                  31
Forza_horizon_4               31
Forza_horizon_5               31
Name: datetime, dtype: int64

we have a daily records for each game. the resampled has worked as expected.

In [183]:
df_resampled.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25356 entries, 0 to 25355
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype        
---  ------           --------------  -----        
 0   game             25356 non-null  object       
 1   datetime         25356 non-null  datetime64[s]
 2   players          24225 non-null  float64      
 3   average_players  24225 non-null  float64      
 4   twitch_viewers   24225 non-null  float64      
 5   month            24225 non-null  float64      
 6   year             24225 non-null  float64      
dtypes: datetime64[s](1), float64(5), object(1)
memory usage: 1.4+ MB


In [184]:
# Checking the null values
print(df_resampled.isnull().sum())

game                  0
datetime              0
players            1131
average_players    1131
twitch_viewers     1131
month              1131
year               1131
dtype: int64


In [185]:
# remove NaN values in columns:players, average_players, twitch_viewers
df_resampled = df_resampled.dropna(subset=['players', 'average_players', 'twitch_viewers'], how='all')
df_resampled.isna().sum()

game               0
datetime           0
players            0
average_players    0
twitch_viewers     0
month              0
year               0
dtype: int64

### Viz of players after downsampling


In [186]:
df_resampled.info()

<class 'pandas.core.frame.DataFrame'>
Index: 24225 entries, 0 to 25355
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype        
---  ------           --------------  -----        
 0   game             24225 non-null  object       
 1   datetime         24225 non-null  datetime64[s]
 2   players          24225 non-null  float64      
 3   average_players  24225 non-null  float64      
 4   twitch_viewers   24225 non-null  float64      
 5   month            24225 non-null  float64      
 6   year             24225 non-null  float64      
dtypes: datetime64[s](1), float64(5), object(1)
memory usage: 1.5+ MB


In [187]:
df_resampled = df_resampled.rename(columns={"index": "date"})
if 'index' in df_resampled.columns:
    df_resampled = df_resampled.drop('index', axis=1)

In [188]:
# bar plot representing count of records bet between 2013 & 2023
hist_new = px.histogram(df_resampled, x='datetime', template='plotly_dark')
hist_new.show()

In [189]:
df_resampled.set_index('datetime')


game  players  average_players  \
datetime                                                         
2015-04-19  American_truck_simulator      1.0              0.0   
2015-04-21  American_truck_simulator      1.0              0.0   
2015-04-25  American_truck_simulator      1.0              0.0   
2015-04-29  American_truck_simulator      1.0              0.0   
2015-05-02  American_truck_simulator      1.0              0.0   
...                              ...      ...              ...   
2023-12-27           Forza_horizon_5  12773.0          16719.0   
2023-12-28           Forza_horizon_5  12726.0          16998.0   
2023-12-29           Forza_horizon_5  13925.0          18034.0   
2023-12-30           Forza_horizon_5  13610.0          18661.0   
2023-12-31           Forza_horizon_5  13948.0          18762.0   

            twitch_viewers  month    year  
datetime                                   
2015-04-19             0.0    4.0  2015.0  
2015-04-21             0.0    4.0  2015.0  
2015-04-25             0.0    4.0  2015.0  
2015-04-29             0.0    4.0  2015.0  
2015-05-02             0.0    5.0  2015.0  
...                    ...    ...     ...  
2023-12-27           420.0   12.0  2023.0  
2023-12-28           397.0   12.0  2023.0  
2023-12-29           737.0   12.0  2023.0  
2023-12-30           463.0   12.0  2023.0  
2023-12-31           395.0   12.0  2023.0  

[24225 rows x 6 columns]

In [190]:
df_resampled.columns

Index(['game', 'datetime', 'players', 'average_players', 'twitch_viewers',
       'month', 'year'],
      dtype='object')

## Ask & Answer questions

### Is sim racing getting more popular ?

In [191]:
# Group data to calculate the average players across all games
global_trend = df_resampled.groupby('datetime')['players'].mean().reset_index()

# Create line chart for global trend
fig = go.Figure()

fig.add_trace(
    go.Scatter(
        x=global_trend['datetime'],
        y=global_trend['players'],
        mode='lines',
        name='Global Trend',
        line=dict(color='purple', width=1)
    )
)

fig.update_layout(
    title='Global trend of average players',
    width=1000,
    height=500,
    yaxis_title='Average players',
    template='plotly_dark',
    xaxis=dict(showgrid=False),
    yaxis=dict(showgrid=False)
)


We see a global positive trend in the last 10 years.
There was a huge peak in December 2014 with 21'620 players on average.
But zooming on the year 2023

### What is the game with the most active players?

In [192]:
df_resampled['game'].unique()

array(['American_truck_simulator', 'Assetto_corsa',
       'Assetto_corsa_competizione', 'Automobilista_2', 'BeamNG',
       'CarX_drift_racing_online', 'Dirt_rally_2.0', 'Euro_truck_2',
       'Forza_horizon_4', 'Forza_horizon_5'], dtype=object)

In [217]:
# Define the game names using a dictionary to remove the undescores
game_names = {
    'American_truck_simulator': 'American truck simulator',
    'Assetto_corsa': 'Assetto corsa',
    'Assetto_corsa_competizione': 'Assetto corsa competizione',
    'Automobilista_2': 'Automobilista 2',
    'BeamNG': 'BeamNG',
    'CarX_drift_racing_online': 'CarX drift racing online',
    'Dirt_rally_2.0': 'Dirt rally 2.0',
    'Euro_truck_2': 'Euro truck 2',
    'Forza_horizon_4': 'Forza horizon 4',
    'Forza_horizon_5': 'Forza horizon 5'
}

In [194]:
df_resampled.info()

<class 'pandas.core.frame.DataFrame'>
Index: 24225 entries, 0 to 25355
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype        
---  ------           --------------  -----        
 0   game             24225 non-null  object       
 1   datetime         24225 non-null  datetime64[s]
 2   players          24225 non-null  float64      
 3   average_players  24225 non-null  float64      
 4   twitch_viewers   24225 non-null  float64      
 5   month            24225 non-null  float64      
 6   year             24225 non-null  float64      
dtypes: datetime64[s](1), float64(5), object(1)
memory usage: 1.5+ MB


In [245]:
# Calculate the average players for each game
avg_daily = round(df_resampled.groupby('game')['average_players'].mean().sort_values(ascending=False))

# Get the game with the most average players
most_avg_daily = avg_daily.idxmax()

# Plot the bar chart for the games with the most average players
bar_avg_daily = px.bar(
    x=avg_daily.index,
    y=avg_daily.values,
    title='Daily average player counts (2013-2023)',
    labels={'x': 'Game', 'y': 'Average Players'},
    color=avg_daily.index,
    color_discrete_map={most_avg_daily: 'red'}
)
bar_avg_daily.update_layout(
    template='plotly_dark',
    showlegend=False
)

bar_avg_daily.update_xaxes(
    labelalias=game_names
)

# Add a horizontal line for the maximum average players
bar_avg_daily.add_hline(
    y=avg_daily.max(),
    line_dash="dash",
    line_color="red",
    annotation_text=f"Max Average Players ({round(avg_daily.max())})",
    annotation_position="top right"
)

# Show the plot
bar_avg_daily.show()


In [239]:
# Game with the most average players
average_players = round(df_resampled.groupby('game')['players'].mean().sort_values(ascending=False))
game_with_most_average_players = average_players.idxmax()

# Plot the bar chart for the game with the most average players
fig_average_players = px.bar(
    x=average_players.index,
    y=average_players.values,
    title='Average player activity (2013-2023)',
    labels={'x': 'Game', 'y': 'Average Players'},
    color_discrete_sequence=["lightgrey"]
)
fig_average_players.update_layout(
    template='plotly_dark')

fig_average_players.add_hline(
    y=average_players.max(),
    line_dash="dash",
    line_color="red",
    annotation_text=f"Max Average Players ({round(average_players.max())})",
    annotation_position="top right")


# Show the plot
fig_average_players.show()


Euro truck 2 is on average the game with the most active players of the last decade (2013-2023), followed by Forza horizon 5 & BeamNG.

### What is the number of players per game over time?

In [ ]:
# Create line chart for the current game
line_chart = px.line(df_resampled, x='datetime', y='players', color='game', title='Daily players per game', template='plotly_dark')

# Add date picker widget with customized button colors using CSS
line_chart.update_layout(
    xaxis=dict(
        showgrid=False,
        title='Move the sliders to adapt the time window. Click on game labels to ',
        rangeselector=dict(
            buttons=list([
                dict(count=1, label="1m", step="month", stepmode="backward"),
                dict(count=6, label="6m", step="month", stepmode="backward"),
                dict(count=1, label="1y", step="year", stepmode="backward"),
                dict(count=3, label="3y", step="year", stepmode="backward"),
                dict(step="all")
            ]),
            bgcolor="rgba(255, 255, 255, 0.5)",  # Background color of the rangeselector
            activecolor="rgba(255, 255, 255, 0.8)",  # Active button color
            bordercolor="rgba(0, 0, 0, 0.2)",
            borderwidth=1,
            font=dict(color="rgba(0, 0, 0, 0.8)")  # Font color
        ),
        rangeslider=dict(visible=True),
        type="date"
    ),
    yaxis=dict(showgrid=False,
                title=''),
    # Hide y-axis labels
    yaxis_tickmode="array",
    yaxis_tickvals=[],
    yaxis_ticktext=[]
)

# Show line chart with all games separately
line_chart.show()


Head-to-head:

- Forza horizon 4 VS Forza horizon 5:
    Despite being the older game, FH 4 keeps a certain popularity among Steam players, even 

In [ ]:
# Extract the day of the week as an integer (0 for Monday, 1 for Tuesday, etc.)
df_resampled['day_of_week'] = df_resampled['datetime'].dt.dayofweek

# Map the integer day of the week to the corresponding weekday name
weekday_names = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']
df_resampled['weekday'] = df_resampled['day_of_week'].map(lambda x: weekday_names[x])
df_resampled.drop(columns=['day_of_week'], inplace=True)
df_resampled

In [ ]:
# avg_players_per_day 
weekdays = round(df_resampled.groupby('weekday').players.mean().sort_values(ascending=False))
weekdays

Sunday is the day when the highest number of players are typically observed.

On average, fewer players are seen on Wednesdays.

In [ ]:
game_counts = df_resampled.groupby('game').size()
print(game_counts)


In [ ]:
df_resampled.columns

In [ ]:
# Create the animated bar chart race
frames = []
for i, date in enumerate(df_resampled['datetime']):
    frame = go.Frame(data=[go.Bar(
        x=df_resampled.columns[1:],
        y=df_resampled.iloc[i, 1:],
        name=str(date)
    )])
    frames.append(frame)

fig = go.Figure(frames=frames)
fig.update_layout(
    title='Bar Chart Race',
    yaxis_title='players',
    template='plotly_dark'
)

# Set animation settings
fig.update_layout(updatemenus=[dict(type='buttons',
                                    showactive=False,
                                    buttons=[dict(label='Play',
                                                  method='animate',
                                                  args=[None, dict(frame=dict(duration=500, redraw=True),
                                                                    fromcurrent=True)]),
                                             dict(label='Pause',
                                                  method='animate',
                                                  args=[[None], dict(frame=dict(duration=0, redraw=True),
                                                                     mode='immediate')])])])

fig.show()

In [ ]:
game_stats = df_resampled.groupby('game').agg(
    {'players':['mean','max'],
     'twitch_viewers':['mean','max']
})
game_stats

### Is there a correlation between average players count & twitch viewers ?

In [ ]:
import seaborn as sns
import matplotlib.pyplot as plt

# Relationship between players & twitch viewers
fig = px.scatter(df_resampled, x="players", y="twitch_viewers", color="game", template='plotly_dark')


# Update layout to remove gridlines
fig.update_layout(
        width=1000,
        height=600,
        xaxis=dict(showgrid=False), 
        yaxis=dict(showgrid=False)
)

# Show the plot
fig.show()

In [ ]:
# Create boxplot with dark template
boxplot = px.box(df_resampled, x='players', template='plotly_dark')

# Update layout to remove gridlines
boxplot.update_layout(
    xaxis=dict(showgrid=False), 
    yaxis=dict(showgrid=False)
)

# Show the plot
boxplot.show()

Median players: 3576.

### K-mean Clustering

Use of K mean clustering to cluster the data

In [ ]:
from sklearn.cluster import KMeans
import matplotlib.pyplot as plt

# Create the KMeans model
kmeans = KMeans(n_clusters=k)

X = df_resampled[['average_players', 'twitch_viewers']]

# Fit the model to the data
kmeans.fit(X)

# Choose the number of clusters (k)
k = 3

# Get the cluster labels for each sample
labels = kmeans.labels_

# Add the cluster labels to the DataFrame
df_resampled['cluster'] = labels

# Visualize the clusters
plt.scatter(df_resampled['players'], df_resampled['average_players'], c=labels, cmap='viridis')
plt.xlabel('Players')
plt.ylabel('Average Players')
plt.title('KMeans Clustering')
plt.colorbar(label='Cluster')
plt.show()


ModuleNotFoundError: No module named 'sklearn'

In [ ]:
# Export to Pickle in data/interim directory
df_resampled.to_pickle('/Users/macbook/Dropbox/Mac/Documents/Pro/Data_Analyst/simracing-players/data/interim/sim_racing_games-1.0.pkl')

## Summary & conclusion

### *Limitations*
- Records come only for steam marketplace
- During our period of analysis we only had 1 record per day for players count.
